README.md

```diff
@@ -1 +1,101 @@
-# edge-block-service
\ No newline at end of file
+# minibus 🚐
+
+> A edge block service to store IPFS data structures
+
+## Getting started
+
+One time set up of your cloudflare worker subdomain for dev:
+
+- `npm install` - Install the project dependencies from the monorepo root directory
+- Sign up to Cloudflare and log in with your default browser.
+- `npm i @cloudflare/wrangler -g` - Install the Cloudflare wrangler CLI
+- `wrangler login` - Authenticate your wrangler cli; it'll open your browser.
+- Copy your cloudflare account id from `wrangler whoami`
+- Update `wrangler.toml` with a new `env`. Set your env name to be the value of `whoami` on your system you can use `pnpm start` to run the worker in dev mode for you.
+
+  [**wrangler.toml**](./wrangler.toml)
+
+  ```toml
+  [env.bobbytables]
+  workers_dev = true
+  account_id = "<what does the `wrangler whoami` say>"
+  ```
+
+- Add secrets
+
+  ```sh
+    wrangler secret put SENTRY_DSN --env $(whoami) # Get from Sentry (not required for dev)
+    wrangler secret put LOGTAIL_TOKEN --env $(whoami) # Get from Logtail
+    wrangler secret put SALT --env $(whoami) # open `https://csprng.xyz/v1/api` in the browser and use the value of `Data`
+  ```
+
+- Add R2 bucket (Note that it is only available as Private Beta at the time of writing)
+
+  ```sh
+  wrangler r2 bucket create blockstore --env $(whoami)
+  # 🌀  Creating bucket "blockstore"
+  # ✨  Success!
+  ```
+
+- `npm run publish` - Publish the worker under your env. An alias for `wrangler publish --env $(whoami)`
+- `npm dev` - Run the worker in dev mode. An alias for `wrangler dev --env $(whoami)`
+
+You only need to `npm dev` for subsequent runs. PR your env config to the `wrangler.toml` to celebrate 🎉
+
+## High level architecture
+
+The Edge HTTP internal API to store IPFS data structures on R2.
+
+![High level Architecture](./minibus.web3.storage-api.jpg)
+
+## Usage
+
+You can see full HTTP API specification at https://web3.storage/block-service-docs.
+
+### 🔒 `PUT /`
+
+> Store individuals blocks by multihash in R2.
+
+TODO: curl example
+
+### 🔒 `GET /:multihash`
+
+> Multihash block read interface.
+
+The block data is retrieved and discovered by multihash, in order to enable the same data to be referred to with different codecs.
```

Comment:
I think it would make sense to allow get by CID as well. Service can just remove codec and lookup by multihash.

Reply:
I like the idea of the redirect! I will get that in

Reply:
redirects cost a round trip. It's a pleasing idea to require a multihash here, but in practice given we can accept a CID and extract the multihash for it I think we should just handle the reques rather than redirect. 

The alternative is we force the client to extract the multihash for every CID or they send us a request and get hit with a 302 and they have to make another request. For popular content and well behaved caches, then maybe it's not a big deal to redirect, but if we have a high percentage of novel CIDs coming through here, then every request becoming 2 requests might not be desirable.

It's probably just a bikeshed... but we should be clear about what we think the benefit of the redirect here is. If we think it's gonna be good for caching, we should check that we don't get all of that benefit by just managing the CID unwrapping internally.

Reply:
like if i use CURL i would have to now have to remember to use a flag to have it follow redirects to get the response i want.

Reply:
Given this service is not to be used outside our services, I am not concerned about extracting multihash extra step. I think the cache benefits of using same URL to request the block will bring benefits to the overall system. The redirect is basically for convenience, but I would expect us to just use multihash

Reply:
im not getting why we need the redirect? is it to cache the multihash as well as the cid ?

Reply:
redirect is to have caching on multihash level

Reply:
we dont need to redirect for that, we can just add both to the cache directly and avoid having to deal with redirects in clients

Reply:
CF Cache is LRU and has limited size. Makes sense to only cache one instance